TFederatedWriteSession: Wait for messages to be sent in Close method

qyryq opened the pull request:
The logic is inspired by ydb_topic's TWriteSessionImpl::Close method

Commit:
TFederatedWriteSession: Wait for messages to be sent in Close method

The logic is inspired by ydb_topic's TWriteSessionImpl::Close method

Commit:
TFederatedWriteSession::Close: wait for the future

Commit:
TFederatedWriteSession: ignore closing in fed state update logic; ignore writes after closing

Commit:
ydb_topic: fix Close method description

Commit:
TFederatedWriteSession: Close

Commit:
TFederatedWriteSession: replace while loop in Close with promises/futures

alexnick88 commented on ydb/public/sdk/cpp/client/ydb_federated_topic/impl/federated_write_session.cpp:
```diff
@@ -27,54 +27,91 @@ NTopic::TTopicClientSettings FromFederated(const TFederatedTopicClientSettings&
+    // A user may have one spare token, so at most one additional message could be written
+    // to the internal queue after the transition to CLOSING or CLOSED state.
+    if (BufferFreeSpace > 0 && !ClientHasToken && SessionState < State::CLOSING) {
+        ClientEventsQueue->PushEvent(NTopic::TWriteSessionEvent::TReadyToAcceptEvent{IssueContinuationToken()});
```
This must be done not under lock. Client's callback may be called here.

alexnick88 commented on ydb/public/sdk/cpp/client/ydb_federated_topic/impl/federated_write_session.cpp:
```diff
@@ -27,54 +27,91 @@ NTopic::TTopicClientSettings FromFederated(const TFederatedTopicClientSettings&
-    AsyncInit.Subscribe([selfCtx = SelfContext](const auto& f){
+    Settings.EventHandlers_.HandlersExecutor_->Start();
+
+    AsyncInit.Subscribe([selfCtx = SelfContext](const auto& f) {
```
This subscribe must be setted not under lock. In may be called right now.

Commit:
TFederatedWriteSession: rename: call *Impl with lock, else w/o lock

E.g. UpdateFederationStateImpl method MUST be called under lock,
but WriteInternal MUST be called without lock.

Commit:
TFederatedWriteSession: add status to log message on Fail to select db

And test that we get here in WriteSessionNoAvailableDatabase test.

Commit:
TFederatedWriteSession: replace START{ING,ED} states with WORKING

Commit:
TFederatedWriteSession: start HandlersExecutor before issuing a token

Add WriteSessionWriteInHandlers test.

Commit:
TFederatedWriteSession: push events to queue without locking
